core/client/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -404,4 +412,75 @@ private void updateBlockInStream(long blockId) throws IOException {
       mShouldCacheCurrentBlock = mAlluxioStorageType.isStore();
     }
   }
+
+  /**
+   * Seeks to a file position. Blocks are not cached unless they are fully read. This is only called
+   * by {@link FileInStream#seek}.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength)
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekInternal(long pos) throws IOException {
+    seekBlockInStream(pos);
+    checkAndAdvanceBlockInStream();
+    mCurrentBlockInStream.seek(mPos % mBlockSize);
+  }
+
+  /**
+   * Seeks to a file position. Blocks are cached even if they are not fully read. This is only
+   * called by {@link FileInStream#seek}.
+   * Invariant: if the current block is to be cached, [0, mPos) should have been cached already.
+   *
+   * @param pos The position to seek to. It is guaranteed to be valid (pos >= 0 && pos != mPos &&
+   *            pos <= mFileLength).
+   * @throws IOException if the seek fails due to an error accessing the stream at the position
+   */
+  private void seekInternalWithCachingIncompleteBlock(long pos) throws IOException {
+    // Precompute this because mPos will be updated several times in this function.
+    boolean isInCurrentBlock = pos / mBlockSize == mPos / mBlockSize;
+
+    if (mShouldCacheCurrentBlock) {
+      // Cache till pos if seeking forward within the current block. Otheriwse cache the whole
+      // block.
+      readCurrentBlockTill(pos > mPos ? pos : mFileLength);
+    }
+
+    // If seeks within the current block, directly seeks to pos if we are not yet there.
+    // If seeks outside the current block, seek to the beginning of that block first, then
+    // cache the prefix (pos % mBlockSize) of that block.
+    if (isInCurrentBlock) {
+      // We don't need to seek if we are at pos already. It won't work if you remove this early
+      // return because SeekBlockInStream closes the current cache stream.
+      if (mPos == pos) {
+        return;
+      }
+
+      seekBlockInStream(pos);
+      mCurrentBlockInStream.seek(mPos % mBlockSize);
+    } else {
+      seekBlockInStream(pos / mBlockSize * mBlockSize);
+      checkAndAdvanceBlockInStream();
+      readCurrentBlockTill(pos);
+    }
+  }
+
+  /**
+   * Reads till the file offset (mPos) equals pos or the end of the current block (whichever is
+   * met first) if pos > mPos. Otherwise no-op.
+   *
+   * @param pos file offset
+   * @throws IOException if read or cache write fails
+   */
+  private void readCurrentBlockTill(long pos) throws IOException {
+    long len = Math.min(pos - mPos,
+        mCurrentBlockInStream == null ? mBlockSize : mCurrentBlockInStream.remaining());
+    if (len <= 0) {
+      return;
+    }
+    Preconditions.checkState(len <= Integer.MAX_VALUE);
+
+    byte[] buffer = new byte[(int) len];
+    read(buffer);
```

Comment:
Since we are discarding the bytes anyway, does it make sense to allocate such a large buffer?